packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -648,13 +648,21 @@ export class RunningSummarizer implements IDisposable {
 						return;
 					}
 
-					// Check for retryDelay that can come from summaryNack or upload summary flow.
-					// Retry the same step only once per retryAfter response.
 					overrideDelaySeconds = result.retryAfterSeconds;
-					if (overrideDelaySeconds === undefined || summaryAttemptsPerPhase > 1) {
+
+					// There are couple of scenarios where we want to retry an attempt with the same options:
+					// 1. If the server asked to retry after a delay and this is the first attempt in a phase. This can happen
+					// due to throttling during summary upload or we got a summaryNack.
+					// 2. If container runtime hits transient failures that may be fixed on a retry. For example, failures such
+					// as local data stores created during summarize can be fixed on retry. This should never happen in summarizer
+					// but unfortunately it does today.
+					const retrySameAttempt =
+						overrideDelaySeconds !== undefined && summaryAttemptsPerPhase === 1;
```

Comment:
>&&  [](http://example.com/codeflow?start=41&length=3)

If this.mc.config.getNumber("Fluid.Summarizer.Attempts")  is 4 we would not retry even when the server sends us result.retryAfterSeconds. 

As for the options, which scenario we would hit  #2 ? 
We still need to increment summaryAttemptPhase, don't we ? 

Shouldn't it  be ? 
if (overrideDelaySeconds === undefined || summaryAttemptsPerPhase >= 1) {

Reply:
I'm confused about this logic too, let's at least get tests illustrating the boundary cases to confirm we got it right.

Reply:
Yeah, this is super confusing. I wanted to rename the different variables to make it clear but that would have increased the changes unncessarily. Let me try to explain the logic and let me know if it clears things up:
- Multiple summaries are attempted and each has a different set of options. Each attempt is called `summaryAttemptPhase`.
- If any attempt fails because of an error that can be retried (retryAfterSeconds is set), the same attempt is tried again with the exact same options. Basically, `summaryAttemptPhase` is not incremented so that the same attempts is retried. `summaryAttemptsPerPhase` (note the `per` added here) is incremented indicating that for a given attempt we are trying again.
  - Note that the above "retry" is done only once per attempt, hence the check for `summaryAttemptsPerPhase === 1`. If the attempt fails again with `retryAfterSeconds` set, `summaryAttemptsPerPhase` will be 2 and we will not retry.
  - So, if there are N attempts in total and every one fails with`retryAfterSeconds`, we actually try to do N*2 summaries.
  - The idea is that this attempt was fine and only failed because of transient error and so, we don't need to change the options.

Reply:
Might be worth a small one-off PR to clean this up, since you've given it so much thought

Reply:
yeah, that's my plan.

Reply:
https://dev.azure.com/fluidframework/internal/_workitems/edit/4708